src/components/Survey/SurveyCourseList.tsx

```diff
@@ -0,0 +1,301 @@
+} from "../../stores/preferences";
+
+const COURSES = gql`
+  query GetCourses {
+    survey {
+      courses {
+        subject
+        code
+        term
+      }
+    }
+  }
+`;
+
+interface CourseOption extends OptionBase {
+  label: string;
+  value: PreferenceInterface;
+}
+
+interface ChildProps {
+  handlePreferenceChange(course: CourseInterface, value: number): void;
+  removeCourse(course: CourseCodeAndSubject): void;
+  removeAllCourses(): void;
+}
+
+export const SurveyCourseList: React.FC<ChildProps> = (props) => {
+  const { loading, error, data } = useQuery(COURSES);
+  const [courseList, setCourseList] = useState<Array<CourseOption>>([]);
+  const [selectedCourses, setSelectedCourses] = useState<Array<CourseOption>>(
+    []
+  );
+
+  const toast = useToast();
+
+  const handleChange = (
+    course: PreferenceInterface,
+    changeType: string,
+    value: string
+  ) => {
+    if (changeType === "Able") {
+      const newSelected = selectedCourses.map(
+        (selectedCourse: CourseOption) => {
+          if (
+            course.code === selectedCourse.value.code &&
+            course.subject === selectedCourse.value.subject
+          ) {
+            props.handlePreferenceChange(
+              {
+                subject: course.subject,
+                code: course.code,
+                term: course.term,
+                preference: 0,
+              },
+              calculateCourseRating(value, selectedCourse.value.willing)
+            );
+            return {
+              ...selectedCourse,
+              value: {
+                ...selectedCourse.value,
+                able: value,
+              },
+            };
+          } else return selectedCourse;
+        }
+      );
+      setSelectedCourses(newSelected);
+    } else {
+      const newSelected = selectedCourses.map(
+        (selectedCourse: CourseOption) => {
+          if (
+            course.code === selectedCourse.value.code &&
+            course.subject === selectedCourse.value.subject
+          ) {
+            props.handlePreferenceChange(
+              {
+                subject: course.subject,
+                code: course.code,
+                term: course.term,
+                preference: 0,
+              },
+              calculateCourseRating(selectedCourse.value.able, value)
+            );
+            return {
+              ...selectedCourse,
+              value: {
+                ...selectedCourse.value,
+                willing: value,
+              },
+            };
+          } else return selectedCourse;
+        }
+      );
+      setSelectedCourses(newSelected);
+    }
+  };
+
+  const containsCourse = (
+    courses: CourseOption[],
+    courseToFind: PreferenceInterface
+  ) => {
+    const filtered = courses.filter((course) => {
+      return (
+        course.value.subject === courseToFind.subject &&
+        course.value.code === courseToFind.code
+      );
+    });
+    return filtered.length > 0;
```

Comment:
You should replace this with this:
```suggestion
    return courses.some(course => 
      couse.value.subject === courseToFind.subject &&
      course.value.code === courseToFind.code
    );
```

Using `filter` here creates an unnecessary copy of the original array

Reply:
that makes sense, thanks